packages/babel-plugin-react-jsx/src/TransformJSXToReactBabelPlugin.js

```diff
@@ -623,25 +643,342 @@ module.exports = function(babel) {
+      });
+
+      if (state.shouldCacheImportFns) {
+        Object.keys(importMap).forEach(importName => {
+          const importPath = getNamedExportImportPath(
+            path,
+            importMap[importName],
+            importName,
+            state.source,
+          );
+
+          const importIdentifier = path.scope.generateUidIdentifier(importName);
+          importPath.insertAfter(
+            t.variableDeclaration('var', [
+              t.variableDeclarator(
+                importIdentifier,
+                t.identifier(importMap[importName]),
+              ),
+            ]),
+          );
+          importMap[importName] = importIdentifier.name;
+        });
+      }
+
+      return importMap;
+    }
+
+    // add import to file and get the import name
+    let name;
+    if (state.autoImport === IMPORT_TYPES.require) {
+      // var _react = require("react");
+      name = addNamespace(path, state.source, {
+        importedInterop: 'uncompiled',
+      }).name;
+    } else if (state.autoImport === IMPORT_TYPES.namespace) {
+      // import * as _react from "react";
+      name = addNamespace(path, state.source).name;
+    } else if (state.autoImport === IMPORT_TYPES.defaultExport) {
+      // import _default from "react";
+      name = addDefault(path, state.source).name;
+    }
+
+    // cache react function names as variables at the top of the file
+    if (state.shouldCacheImportFns) {
+      // var _jsx = _react.jsx;
+      // var _createElement = _react.createElement;
+      const imports = getImportNames(path, state);
+      let importPath;
+      let importMap = {};
+      if (state.autoImport === IMPORT_TYPES.require) {
+        importPath = getRequirePath(path, name, state.source);
+      } else if (state.autoImport === IMPORT_TYPES.namespace) {
+        importPath = getNamespaceImportPath(path, name, state.source);
+      } else if (state.autoImport === IMPORT_TYPES.defaultExport) {
+        importPath = getDefaultImportPath(path, name, state.source);
+      }
+
+      Object.keys(imports).forEach(importName => {
+        const importIdentifier = path.scope.generateUidIdentifier(importName);
+        importPath.insertAfter(
+          t.variableDeclaration('var', [
+            t.variableDeclarator(
+              importIdentifier,
+              t.memberExpression(t.identifier(name), t.identifier(importName)),
+            ),
+          ]),
+        );
+        importMap[importName] = importIdentifier.name;
+      });
+      return importMap;
+    } else {
+      // don't cache react function names
+      return name;
+    }
+  }
+
   visitor.Program = {
     enter(path, state) {
+      let autoImport = state.opts.autoImport || 'none';
+      let source = state.opts.importSource || 'react';
+      const CACHE_IMPORT_FNS = state.opts.shouldCacheImportFns || false;
+      const {file} = state;
+
+      if (file.ast.comments) {
+        for (const comment of (file.ast.comments: Array<Object>)) {
```

Comment:
This is not a Flow file. Remove the annotation.
Also, let's just use `for (let i = 0; ...` loop to avoid the lint problem.